### PR TITLE
[server][wfs] Apply BBOX inside And using filterRect

### PR DESCRIPTION
**Pull request by arnaud-morvan**
In case of a WFS GetFeature request (or WMS GetMap request with FILTER),
if BBOX is not a direct child of the Filter element,
it is applyed through an intersects_bbox function in the QgsFeatureRequest filterExpression.
This is not compiled by providers like PostgreSQL, causing the whole filter to be interpreted on QGIS side.
When interpreted on QGIS side, the srsname given in the request is not handled properly as geom_from_gml return a geometry object, projection agnostic.
This result in a very long request returning no results.

This is a workaround for this performance and srs issue in the case the BBOX is direct child of an And operator, itself at first level in Filter element.

This is a bug fix and huge optimisation for the case we have an And with a BBOX and another condition.

Note that it may be possible to transmit intersects_bbox as an && to PostgreSQL when compiling expressions, but it does not come without complexity as && is not a function but an operator and Postgis ST_GeomFromGML does not handle Envelope tag.

I've also tried keeping the filter handling on QGIS side, but use of geom_from_gml does not permit to handle the srsname in Envelope or Query elements.

## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/server/services/wfs/qgswfsutils.cpp
Comment (m-kuhn) [inline src/server/services/wfs/qgswfsutils.cpp]: does capitalization matter?

File: src/server/services/wfs/qgswfsutils.cpp
Comment (m-kuhn) [inline src/server/services/wfs/qgswfsutils.cpp]: Are there potentially more than two child elements and the filter expression is overwritten repeatedly? (disclaimer: I've got no idea and do not know this code)

**Comment by Gustry** [comment]
This PR is for 3.0 which has reached its EOL

File: src/server/services/wfs/qgswfsutils.cpp
Comment (arnaud-morvan) [reply src/server/services/wfs/qgswfsutils.cpp]: Exact, the standard says : 
```<xsd:choice minOccurs="2" maxOccurs="unbounded">```
Multiple expressions need to be wrapped in an "and" expression node.

**Comment by arnaud-morvan** [comment]
@gustry : I've opened this PR before 3.0.3, I will now rebase it over 3.2.

File: src/server/services/wfs/qgswfsutils.cpp
Comment (m-kuhn) [reply src/server/services/wfs/qgswfsutils.cpp]: so the risk of multiple expressions colliding is a real risk?

File: src/server/services/wfs/qgswfsutils.cpp
Comment (arnaud-morvan) [reply src/server/services/wfs/qgswfsutils.cpp]: The standard says : 
```
<xsd:element name="And"
type="fes:BinaryLogicOpType"
substitutionGroup="fes:logicOps"/>
```
As far as "BBOX" is tested with case sensitive matching, I've done the same for "And".

Comment (arnaud-morvan) [reply src/server/services/wfs/qgswfsutils.cpp]: Note that in QgsOgcUtils, "And" is also tested with case sensitive matching.
https://github.com/qgis/QGIS/blob/9cf6bbf42490da69d7513f5de2c6e45e9a8ab652/src/core/qgsogcutils.cpp#L1654:L1673

File: src/server/services/wfs/qgswfsutils.cpp
Comment (m-kuhn) [inline src/server/services/wfs/qgswfsutils.cpp]: memory leak?

**Comment by Gustry** [comment]
I didn't know we could change the base branch after the PR was created :)

File: src/server/services/wfs/qgswfsutils.cpp
Comment (arnaud-morvan) [reply src/server/services/wfs/qgswfsutils.cpp]: I'm not completely aware with memory leaks but, after some debugging, it seems that deleting `node` also delete all child nodes, and other expressions and related nodes are also deleted with QgsFeatureRequest object. So it should be ok now with `delete node`.

Comment (m-kuhn) [reply src/server/services/wfs/qgswfsutils.cpp]: Yep, looks better now.
It would be prefect if you could use
`std::unique_ptr<QgsExpressionNodeBinaryOperator> node = std::make_unique<QgsExpressionNodeBinaryOperator>( ... );`

Comment (m-kuhn) [reply src/server/services/wfs/qgswfsutils.cpp]: Thanks for the update.
Did the `make_unique` not work?

Comment (arnaud-morvan) [reply src/server/services/wfs/qgswfsutils.cpp]: I missed the make_unique, done with qgis::make_unique.

**Comment by arnaud-morvan** [comment]
@gustry : I think this is not old in GitHub, click on PR title edit button, base branch change for a dropdown.

**Comment by arnaud-morvan** [comment]
How about merging this ? I do not have push right.

**Comment by rldhont** [comment]
@arnaud-morvan If @m-kuhn validates it, I ill merge it.

Do you think this has to be backported to release-2_18 ?

**Review by rldhont — APPROVED** [review]

File: src/server/services/wfs/qgswfsutils.cpp
Comment (m-kuhn) [inline src/server/services/wfs/qgswfsutils.cpp]: Can you rewrite this to `request.filterExpression().isEmpty()`?

Comment (m-kuhn) [reply src/server/services/wfs/qgswfsutils.cpp]: Ah wait, this returns a `QgsExpression*`, not a string. Just drop the `bool()` in this case.

**Comment by arnaud-morvan** [comment]
@rldhont : As it fix an issue with SRS and an huge performance issue in specific cases, it would make sense to backport to 2.18. I will create the branch for the 2.18, and let you known when it's done.

**Review by m-kuhn — APPROVED** [review]

**Comment by rldhont** [comment]
Thanks for backporting.
I am waiting for travis before merging.

**Comment by arnaud-morvan** [comment]
Tried to cherry-pick this commit in release-2_18. As server source files has been completely refactorized, it seems a little complex to backport this to release-2_18. I would prefer spending time on improving QGIS server 3.

**Comment by rldhont** [comment]
I will manually cherry-pick it!

**Comment by rldhont** [comment]
@arnaud-morvan does this PR fixe the issue https://issues.qgis.org/issues/19100 WFS request with combined filter (BBOX + PropertyIsEqual) not working

**Comment by arnaud-morvan** [comment]
Yes, this PR was fixing this specific case (BBOX is direct child of And, itself direct child of Filter element).
Note that the bug still exists if BBOX is not direct child of Filter or direct child of an And itself direct child of Filter.
For example, if BBOX is inside an OR, BBOX will be interpreted as an QGIS expression looking like : 
`intersects_bbox(geom_from_gml('Envelope(...` and intersects_bbox is not compiled to postgis, so it is interpreted on QGIS side (after fectching all records) and QGIS does not handle the Envelope SRS attribute, as geom_from_gml return a geometry SRS agnostic.
In such case, GetFeature returns no results.

Also note that postgis ST_GeomFromGML does not handle the Envelope GML tag.

**Pull request merged**
